apps/rush-lib/src/logic/operations/ShellOperationRunnerPlugin.ts

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import type { IPhase } from '../../api/CommandLineConfiguration';
+import type { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import { RushConstants } from '../RushConstants';
+import { NullOperationRunner } from './NullOperationRunner';
+import { convertSlashesForWindows, ShellOperationRunner } from './ShellOperationRunner';
+import { OperationStatus } from './OperationStatus';
+import type { ICreateOperationsContext, PhasedCommandHooks } from '../../pluginFramework/PhasedCommandHooks';
+import { Operation } from './Operation';
+
+const PLUGIN_NAME: 'ShellOperationRunnerPlugin' = 'ShellOperationRunnerPlugin';
+
+/**
+ * Core phased command plugin that provides the functionality for executing an operation via shell command.
+ */
+export class ShellOperationRunnerPlugin {
+  public apply(hooks: PhasedCommandHooks): void {
+    hooks.createOperations.tap(PLUGIN_NAME, createShellOperations);
+
+    function createShellOperations(
+      operations: Set<Operation>,
+      context: ICreateOperationsContext
+    ): Set<Operation> {
+      const {
+        buildCacheConfiguration,
+        customParameters,
+        isIncrementalBuildAllowed,
+        phaseSelection: selectedPhases,
+        projectChangeAnalyzer,
+        rushConfiguration
+      } = context;
+
+      const customParametersByPhase: Map<IPhase, string[]> = new Map();
+
+      function getCustomParameterValuesForPhase(phase: IPhase): ReadonlyArray<string> {
+        let customParameterValues: string[] | undefined = customParametersByPhase.get(phase);
+        if (!customParameterValues) {
+          customParameterValues = [];
+          for (const [parameter, tsCommandLineParameter] of customParameters) {
```

Comment:
Flip the loop and the if to reduce the total amount of computation. May also investigate directly attaching the resolved CommandLineParameter objects to the phase.